taroscript/send.go

```diff
@@ -0,0 +1,232 @@
+package taroscript
+
+import (
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/lightninglabs/taro/asset"
+	"github.com/lightninglabs/taro/commitment"
+	"github.com/lightninglabs/taro/taropsbt"
+)
+
+var (
+	// ErrInvalidCollectibleSplit is returned when a collectible is split
+	// into more than two outputs.
+	ErrInvalidCollectibleSplit = errors.New(
+		"fund: invalid collectible split",
+	)
+
+	// ErrInvalidChangeOutputLocation is returned when the change output is
+	// not at the expected location (index 0).
+	ErrInvalidChangeOutputLocation = errors.New(
+		"fund: invalid change output location, should be index 0",
+	)
+
+	// ErrInvalidSplitAmounts is returned when the split amounts don't add
+	// up to the amount of the asset being spent.
+	ErrInvalidSplitAmounts = errors.New(
+		"fund: invalid split amounts, sum doesn't match input",
+	)
+)
+
+// PrepareOutputAssets prepares the assets of the given outputs depending on
+// the amounts set on the transaction. If a split is necessary (non-interactive
+// or partial amount send) it computes a split commitment with the given inputs
+// and spend information. The inputs MUST be checked as valid beforehand and the
+// change output is expected to be declared as such (and be at index 0).
+func PrepareOutputAssets(vPkt *taropsbt.VPacket) error {
+	// We currently only support a single input.
+	//
+	// TODO(guggero): Support multiple inputs.
+	if len(vPkt.Inputs) != 1 {
+		return fmt.Errorf("only a single input is currently supported")
+	}
+	input := vPkt.Inputs[0]
+	outputs := vPkt.Outputs
+
+	// This should be caught way earlier but just to make sure that we never
+	// overflow when converting the input amount to int64 we check this
+	// again.
+	inputAsset := input.Asset()
+	if inputAsset.Amount > math.MaxInt64 {
+		return fmt.Errorf("amount int64 overflow")
+	}
+
+	// Do some general sanity checks on the outputs, these should be
+	// independent of the number of outputs.
+	for idx := range outputs {
+		scriptPubKey := outputs[idx].ScriptKey.PubKey
+
+		// We need at least the script key set on each output.
+		if scriptPubKey == nil {
+			return fmt.Errorf("output %d has no script key", idx)
+		}
+
+		// Any non change output shouldn't send to an un-spendable
+		// script, burning assets is done in a different way.
+		if !outputs[idx].IsChange &&
+			scriptPubKey.IsEqual(asset.NUMSPubKey) {
+
+			return commitment.ErrInvalidScriptKey
+		}
+	}
+
+	switch {
+	// We need at least one output.
+	case len(outputs) == 0:
+		return fmt.Errorf("no outputs specified in virtual packet")
+
+	// A single output implies an interactive send. The value should be
+	// equal to the input amount and the script key should be a spendable
+	// one.
+	case len(outputs) == 1:
+		vOut := outputs[0]
+
+		if !vOut.Interactive {
+			return fmt.Errorf("single output must be interactive")
+		}
+
+		if vOut.Amount != inputAsset.Amount {
+			return ErrInvalidSplitAmounts
+		}
+
+		if vOut.ScriptKey.PubKey.IsEqual(asset.NUMSPubKey) {
```

Comment:
I think this is ok for now, but we might lift restriction later once we implement provable burns (#24).